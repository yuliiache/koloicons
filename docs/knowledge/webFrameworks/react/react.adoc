== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







