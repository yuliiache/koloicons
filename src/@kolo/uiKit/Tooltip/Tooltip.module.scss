@import "src/@kolo/design";

.tooltipWrapper {
  position: relative;
  display: inline-block;
  font-size: 0;
  outline: none;
  border: none;

  & .tooltip {
    font-size: 14px;
    line-height: 1.43;
    padding: 8px;
    border-radius: 6px;
    color: $WhiteBG;
    z-index: 1;
    background-color: $TooltipBG;
    box-shadow: 0 0 4px rgba(0, 0, 0, 0.17);
  }
}
