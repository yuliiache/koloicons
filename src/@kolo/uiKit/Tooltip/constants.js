export const TOOLTIP_PLACEMENT = Object.freeze({
  LEFT: 'left',
  RIGHT: 'right',
  TOP: 'top',
  BOTTOM: 'bottom',
});

export const TRIGGER_ACTION = Object.freeze({
  HOVER: 'hover',
  FOCUS: 'focus',
  CLICK: 'click',
});
