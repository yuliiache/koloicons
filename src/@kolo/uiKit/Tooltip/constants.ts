export const TOOLTIP_TEST_ID = 'tooltipTestId';

export enum TooltipPlacement {
  LEFT = 'left',
  LEFT_START = 'left-start',
  LEFT_END = 'left-end',
  RIGHT = 'right',
  RIGHT_START = 'right-start',
  RIGHT_END = 'right-end',
  TOP = 'top',
  TOP_START = 'top-start',
  TOP_END = 'top-end',
  BOTTOM = 'bottom',
  BOTTOM_START = 'bottom-start',
  BOTTOM_END = 'bottom-end',
}

export const OPEN_ACTION_TRIGGER = 'mouseenter';
export const CLOSE_ACTION_TRIGGER = 'mouseleave';
