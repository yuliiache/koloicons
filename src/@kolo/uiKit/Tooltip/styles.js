export const overlayInnerStyle = Object.freeze({
  padding: '8px 10px',
  color: '#fff',
  textAlign: 'left',
  textDecoration: 'none',
  backgroundColor: '#373737',
  borderRadius: '6px',
  boxShadow: '0 0 4px rgba(0, 0, 0, 0.17)',
  minHeight: '34px',
  border: 'none',
});
