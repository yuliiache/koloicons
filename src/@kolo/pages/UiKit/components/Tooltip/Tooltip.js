import React from 'react';
import PropTypes from 'prop-types';
import RcTooltip from 'rc-tooltip';

import {TOOLTIP_PLACEMENT, TRIGGER_ACTION} from './constants';
import {overlayInnerStyle} from './styles';

import 'rc-tooltip/assets/bootstrap_white.css';

export const Tooltip = ({label, placement = TOOLTIP_PLACEMENT.LEFT, trigger = TRIGGER_ACTION.HOVER, children}) => {
  if (!label) {
    return children;
  }

  return (
    <RcTooltip
      placement={placement}
      overlay={label}
      trigger={trigger}
      overlayInnerStyle={{...overlayInnerStyle}}
    >
      {children}
    </RcTooltip>
  );
};

Tooltip.propTypes = {
  label: PropTypes.string,
  placement: PropTypes.oneOf(Object.values(TOOLTIP_PLACEMENT)),
  trigger: PropTypes.oneOf(Object.values(TRIGGER_ACTION)),
  children: PropTypes.node,
};
