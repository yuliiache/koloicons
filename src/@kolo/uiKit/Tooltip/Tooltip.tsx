import {FC, ReactNode} from 'react';
import {Tooltip as ReactTooltip} from 'react-tooltip';

import {CLOSE_ACTION_TRIGGER, OPEN_ACTION_TRIGGER, TOOLTIP_TEST_ID, TooltipPlacement} from './constants';
import {createEventObject} from './helpers/createEventObject';
import styles from './Tooltip.module.scss';

interface Props {
  children: ReactNode;
  id?: string;
  label?: string;
  openTrigger?: string;
  closeTrigger?: string;
  placement?: TooltipPlacement;
}

// id and label are required for rendering the Tooltip
const Tooltip: FC<Props> = ({
  id,
  label,
  children,
  openTrigger = OPEN_ACTION_TRIGGER,
  closeTrigger = CLOSE_ACTION_TRIGGER,
  placement = TooltipPlacement.LEFT,
}) => {
  const tooltipId = `tooltip-${id}`;

  if (!label && !id) {
    return <>{children}</>;
  }

  return (
    <div
      data-testid={TOOLTIP_TEST_ID}
      id={tooltipId}
      className={styles.tooltipWrapper}
    >
      {children}
      <ReactTooltip
        anchorSelect={`#${tooltipId}`}
        className={styles.tooltip}
        content={label}
        place={placement}
        openEvents={createEventObject(openTrigger)}
        closeEvents={createEventObject(closeTrigger)}
      />
    </div>
  );
};

export default Tooltip;
