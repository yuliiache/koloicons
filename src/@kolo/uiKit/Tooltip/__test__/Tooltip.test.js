import React from 'react';
import {cleanup, render, screen} from '@testing-library/react';

import Tooltip from '../Tooltip';

describe('Tooltip', () => {
  afterEach(cleanup);

  const label = 'Tooltip';
  const buttonText = 'Test button';

  it(`should render snapshot successfully with label`, () => {
    const {container} = render(
      <Tooltip
        id="1"
        label={label}
      >
        <button>{buttonText}</button>
      </Tooltip>
    );

    const triggerElement = screen.getByText(buttonText);

    expect(triggerElement).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render snapshot successfully for only child element without tooltip', () => {
    const {asFragment} = render(<Tooltip>{<button>{buttonText}</button>}</Tooltip>);

    const triggerElement = screen.getByText(buttonText);

    expect(triggerElement).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
