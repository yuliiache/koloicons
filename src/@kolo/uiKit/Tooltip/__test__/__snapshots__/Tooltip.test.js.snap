// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render snapshot successfully for only child element without tooltip 1`] = `
<DocumentFragment>
  <button>
    Test button
  </button>
</DocumentFragment>
`;

exports[`Tooltip should render snapshot successfully with label 1`] = `
<div>
  <div
    class="tooltipWrapper"
    data-testid="tooltipTestId"
    id="tooltip-1"
  >
    <button>
      Test button
    </button>
  </div>
</div>
`;
